contracts/governance/Governor.sol

```diff
@@ -602,29 +595,54 @@ abstract contract Governor is Context, ERC165, EIP712, Nonces, IGovernor, IERC72
         bytes memory params,
         bytes memory signature
     ) public virtual returns (uint256) {
-        bool valid = SignatureChecker.isValidSignatureNow(
+        bytes memory rawSignatureDigestData = abi.encode(
+            EXTENDED_BALLOT_TYPEHASH,
+            proposalId,
+            support,
             voter,
-            _hashTypedDataV4(
-                keccak256(
-                    abi.encode(
-                        EXTENDED_BALLOT_TYPEHASH,
-                        proposalId,
-                        support,
-                        voter,
-                        _useNonce(voter),
-                        keccak256(bytes(reason)),
-                        keccak256(params)
-                    )
-                )
-            ),
-            signature
+            nonces(voter),
+            keccak256(bytes(reason)),
+            keccak256(params)
         );
+        _validateVoteSignature(voter, proposalId, signature, rawSignatureDigestData);
+
+        return _castVote(proposalId, voter, support, reason, params);
+    }
+
+    function _validateVoteSignature(
+        address voter,
+        uint256 proposalId,
+        bytes memory signature,
+        bytes memory rawSignatureDigestData
+    ) internal virtual {
+        bool valid;
+        if (
+            SignatureChecker.isValidSignatureNow(voter, _hashTypedDataV4(keccak256(rawSignatureDigestData)), signature)
+        ) {
+            valid = true;
+            _useNonce(voter);
+        } else {
+            // uint192 is sufficient entropy for proposalId within nonce keys.
+            uint256 keyedNonce = nonces(voter, uint192(proposalId));
+            assembly ("memory-safe") {
+                mstore(add(rawSignatureDigestData, 0xA0), keyedNonce)
+            }
```

Comment:
I'm really not a fan of the assumption made here.

This is an internal function. It could be overriden, of called with different parameters.

Reply:
Would you be ok with making it private?